src/main/java/io/aiven/kafka/connect/http/sender/OAuth2HttpSender.java

```diff
@@ -19,82 +19,98 @@
 import java.io.IOException;
 import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.Builder;
 import java.net.http.HttpResponse;
 import java.util.Map;
 
 import org.apache.kafka.connect.errors.ConnectException;
 
 import io.aiven.kafka.connect.http.config.HttpSinkConfig;
+import io.aiven.kafka.connect.http.sender.DefaultHttpSender.DefaultHttpRequestBuilder;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-final class OAuth2HttpSender extends HttpSender {
+class OAuth2HttpSender extends AbstractHttpSender implements HttpSender {
 
-    private final Logger logger = LoggerFactory.getLogger(OAuth2HttpSender.class);
-
-    private String accessTokenAuthHeader;
-
-    private final ObjectMapper objectMapper = new ObjectMapper();
-
-    OAuth2HttpSender(final HttpSinkConfig config) {
-        super(config);
-    }
-
-    //for testing
-    OAuth2HttpSender(final HttpSinkConfig config, final HttpClient httpClient) {
-        super(config, httpClient);
+    OAuth2HttpSender(
+        final HttpSinkConfig config, final HttpClient httpClient, final AccessTokenHttpSender accessTokenHttpSender
+    ) {
+        super(config, new OAuth2AuthHttpRequestBuilder(config, accessTokenHttpSender), httpClient);
     }
 
     @Override
-    protected HttpResponse<String> sendWithRetries(final HttpRequest.Builder requestBuilder,
-                                                   final HttpResponseHandler originHttpResponseHandler) {
-        final var accessTokenAwareRequestBuilder = loadOrRefreshAccessToken(requestBuilder);
-        final HttpResponseHandler handler = response -> {
-            if (response.statusCode() == 401) { // access denied or refresh of a token is needed
-                this.accessTokenAuthHeader = null;
-                this.sendWithRetries(requestBuilder, originHttpResponseHandler);
+    protected HttpResponse<String> sendWithRetries(
+        final Builder requestBuilderWithPayload,
+        final HttpResponseHandler originHttpResponseHandler,
+        final int retriesNumber
+    ) {
+        final HttpResponseHandler handler = (response, retriesRemain) -> {
+            if (response.statusCode() == 401 && retriesRemain > 0) { // access denied or refresh of a token is needed
+                ((OAuth2AuthHttpRequestBuilder) this.httpRequestBuilder).renewAccessToken(requestBuilderWithPayload);
+                this.sendWithRetries(requestBuilderWithPayload, originHttpResponseHandler, retriesRemain - 1);
```

Comment:
Maybe you didn't see this, but actually we do decrease the counter to avoid entering in a loop:
`this.sendWithRetries(requestBuilderWithPayload, originHttpResponseHandler, retriesRemain - 1)`